src/Parallel/PhaseControl/PhaseChange.hpp

```diff
@@ -0,0 +1,212 @@
+ *   of components that contribute data during this reduction. This can be used
+ *   to screen out components that will not have the necessary information to
+ *   contribute to the reduction. If all components should participate, this
+ *   type alias can be set to simply `typename Metavariables::component_list`.
+ *
+ * And member functions with signatures:
+ *
+ * ```
+ * template <typename... DecisionTags>
+ * void initialize_phase_change_decision_data_impl(
+ *     const gsl::not_null<tuples::TaggedTuple<DecisionTags...>*>
+ *         phase_change_decision_data) const noexcept;
+ * ```
+ * - Must set all tags in `phase_change_tags_and_combines` to useful initial
+ *   states in the `phase_change_decision_data`.
+ *
+ * ```
+ * template <typename ParallelComponent, typename ArrayIndex>
+ * void contribtue_phase_data_impl(
+ *     [DataBox return tags...], [DataBox argument tags...],
+ *     Parallel::GlobalCache<Metavariables>& cache,
+ *     const ArrayIndex& array_index) const noexcept;
+ * ```
+ * - Should send any data relevant for the associated phase change decision made
+ *   in `arbitrate_phase_control_impl` to the Main chare via function
+ *   `Parallel::contribute_to_phase_change_reduction`.
+ *
+ * ```
+ * template <typename... DecisionTags, typename Metavariables>
+ * typename std::optional<
+ *     std::pair<typename Metavariables::Phase, ArbitrationStrategy>>
+ * arbitrate_phase_control_impl(
+ *     const gsl::not_null<tuples::TaggedTuple<DecisionTags...>*>
+ *         phase_change_decision_data,
+ *     const typename Metavariables::Phase current_phase,
+ *     const Parallel::GlobalCache<Metavariables>& cache) const noexcept;
+ * ```
+ * - Should examine the collected data in `phase_change_decision_data` and
+ *   optionally return a `std::pair` with the desired `Metavariables::Phase` and
+ *   an `PhaseControl::ArbitrationStrategy` indicating a method for arbitrating
+ *   multiple simultaneous requests. Alternatively, it may return `std::nullopt`
+ *   to abstain from the phase decision.
+ */
+template <typename PhaseChangeRegistrars>
+struct PhaseChange : public PUP::able {
+ protected:
+  /// \cond
+  PhaseChange() = default;
+  PhaseChange(const PhaseChange&) = default;
+  PhaseChange(PhaseChange&&) = default;
+  PhaseChange& operator=(const PhaseChange&) = default;
+  PhaseChange& operator=(PhaseChange&&) = default;
+  /// \endcond
+
+ public:
+  ~PhaseChange() override = default;
+
+  WRAPPED_PUPable_abstract(PhaseChange);  // NOLINT
+
+  using creatable_classes = Registration::registrants<PhaseChangeRegistrars>;
+
+  /// Send data from all `participating_components` to the Main chare for
+  /// determining the next phase.
+  template <typename ParallelComponent, typename DbTags, typename Metavariables,
+            typename ArrayIndex>
+  void contribute_phase_data(const gsl::not_null<db::DataBox<DbTags>*> box,
+                             Parallel::GlobalCache<Metavariables>& cache,
```

Comment:
In this case, it needs to get `CProxy`s to contribute to the reduction, which if I understand correctly, requires access to the cache as a non-const -- I suppose I could put it inside a `db::mutate`, but that starts to seem awkward. If the simplification of the interface seems worth it, though, I could make that change.

Reply:
the GlobalCache also currently holds the components which are not available through the DataBox

Reply:
DataBox actually holds a `Parallel::GlobalCache<Metavariables>*`, so you can get a mutable one using a plain `db::get`.

Reply:
My original intention was that you should not be retrieving that pointer ever, only accessing tags from the cache. The reason for it was that if something does communication it should access the cache explicitly. My worry was/is that if folks start grabbing the cache from the DataBox everywhere we'll end up in the situation where it's difficult to tell what does communication and what doesn't. But if the decision is that passing around the cache is worse than the implicit communications (which can happen silently right now anyway), that's fine. I'm just a bit hesitant on that

Reply:
OK, that's a reasonable distinction.  (Probably should be added to the documentation.)

My annoyance with this has been cases where one derived class needed the cache, so everything needed to use it, even if not needed.  Looking, though, all the cases of this I can find are relics from before cache tags were in the DataBox and are no longer necessary anyway.

The one thing I would like (and why I happened to be thinking about this) is a way to get the Metavariables struct from the DataBox.  The GlobalCache has it as a type alias, which was convenient, but I see the argument against using that.  A compute tag giving a `Metavariables*` would probably be doable.

Reply:
It should be straightforward to add a Metavariables object as a simple item at the same time as the GlobalCache pointer is added as a simple item.

Reply:
I like the idea of having a `Metavariables*` simple tag. I think a pointer should be sufficient (and have the pointer be `nullptr`) since we don't actually use the metavariables as an object.

Reply:
(My original suggestion for `Metavariables*` was because `Metavariables` is not serializable, but of course `Metavariables*` isn't either, so that doesn't help.)